### PR TITLE
[gilbert] Round-2: tangential wall-shear projection on 6L base

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The tangential wall-shear projection loss (PR #11 kohaku, `--use-tangential-wallshear-loss`)
is a physical constraint that drives predicted wall-shear vectors onto the tangential plane,
but it was tested on an older 4L/256d base without gradient clipping. With the new 6L/256d
base (PR #14, abupt=13.15) and gradient clipping now standard, the composition should be
stable and potentially yield further improvement on the wall-shear axes (τ_y=16.23 and
τ_z=16.75 are still 4.5× the AB-UPT targets).

PR #9 (gilbert's protocol fixes) deliberately excluded `--use-tangential-wallshear-loss`
and still won — but the two are orthogonal. Now that depth is providing capacity and
clipping provides stability, the projection constraint may push the wall-shear axes further.

This is a **flag-only, no-code experiment** — the projection code is already on yi.

## Instructions

No code changes needed.

**2-arm sweep (test both with and without the projection to isolate its delta):**

| GPU | `--use-tangential-wallshear-loss` | `--wandb-name` |
|---|---|---|
| 0 | absent (control) | `gilbert-6l-noproj` |
| 1 | present | `gilbert-6l-proj` |

**Run command (template — add/remove `--use-tangential-wallshear-loss` per arm):**

```bash
cd target/
python train.py \
  [--use-tangential-wallshear-loss] \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --volume-loss-weight 2.0 \
  --batch-size 8 \
  --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --clip-grad-norm 1.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group gilbert-6l-tangential-sweep \
  --wandb-name gilbert-6l-<noproj or proj>
```

**Key metrics to watch:**
- `test_primary/wall_shear_y_rel_l2_pct` and `test_primary/wall_shear_z_rel_l2_pct`
  — these are the target axes for the projection loss.
- `train/wallshear_pred_normal_rms` — should decrease when projection is on.
- Does the projection cause train→val divergence at epoch 3+ on the 6L base?
  (Kohaku PR #21 saw this on 4L base without clip; now we have clip=1.0.)

## Baseline (current yi best — PR #14 senku 6L/256d)

| Metric | 6L yi best (`et4ajeqj`) | AB-UPT |
|---|---:|---:|
| `test_primary/abupt_axis_mean_rel_l2_pct` | **13.15** | — |
| `test_primary/surface_pressure_rel_l2_pct` | **7.64** | 3.82 |
| `test_primary/wall_shear_rel_l2_pct` | **13.47** | 7.29 |
| `test_primary/volume_pressure_rel_l2_pct` | **13.58** | 6.08 |
| `test_primary/wall_shear_x_rel_l2_pct` | **11.53** | 5.35 |
| `test_primary/wall_shear_y_rel_l2_pct` | **16.23** | 3.65 |
| `test_primary/wall_shear_z_rel_l2_pct` | **16.75** | 3.63 |

Wall-shear y/z are the largest remaining gap (4.4× and 4.6× AB-UPT).
Any reduction in `wall_shear_y_rel_l2_pct` or `wall_shear_z_rel_l2_pct` is high-value.

## Results

Add a PR comment with:
1. Both arms: all `test_primary/*_rel_l2_pct` and `full_val_primary/*`.
2. Per-epoch trajectory for both arms — look for divergence at epoch 3+.
3. `train/wallshear_pred_normal_rms` trajectory for the projection arm.
4. W&B run IDs and URLs.
5. Verdict: does projection help on the 6L base? Which axes moved most?

## Constraints

- Do **not** modify `data/*`, `pyproject.toml`, or `instructions/*`.
- Do **not** override `SENPAI_MAX_EPOCHS` or `SENPAI_TIMEOUT_MINUTES`.
- `test_primary/*` must **not** be NaN.

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT gilbert:

## Results

Both arms ran for 3 full epochs (timeout-bounded at 272.4 min, SENPAI_TIMEOUT_MINUTES=360 was not exceeded — the runs hit the 4 hr 32 min budget mid-epoch 4 and stopped on the train-timeout path). Best checkpoint was epoch 3 in both cases.

### `test_primary/*` (paper-facing, after best-checkpoint reload)

| Metric | noproj (control) | proj | Δ proj − noproj | AB-UPT | yi best (PR #14) |
|---|---:|---:|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **14.28** | 34.07 | **+19.80** ❌ | — | 13.15 |
| `surface_pressure_rel_l2_pct` (`p_s`) | 8.49 | **7.75** | −0.74 ✅ | 3.82 | 7.64 |
| `wall_shear_rel_l2_pct` (`τ`) | **14.65** | 42.28 | **+27.63** ❌ | 7.29 | 13.47 |
| `wall_shear_x_rel_l2_pct` | **12.42** | 28.09 | +15.67 ❌ | 5.35 | 11.53 |
| `wall_shear_y_rel_l2_pct` | **17.67** | 40.20 | +22.54 ❌ | 3.65 | 16.23 |
| `wall_shear_z_rel_l2_pct` | **18.54** | 80.73 | +62.19 ❌ | 3.63 | 16.75 |
| `volume_pressure_rel_l2_pct` (`p_v`) | 14.26 | **13.60** | −0.67 ✅ | 6.08 | 13.58 |
| `surface_pressure_mae` | 0.0229 | **0.0214** | −0.0016 | — | — |
| `wall_shear_mae` | **0.1336** | 0.3447 | +0.2111 ❌ | — | — |
| `volume_pressure_mae` | 30.717 | **29.397** | −1.32 | — | — |

### `full_val_primary/*` (best-checkpoint full-fidelity validation)

| Metric | noproj | proj |
|---|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **13.0946** | 33.2077 |
| `surface_pressure_rel_l2_pct` | 8.59 | **7.83** |
| `wall_shear_rel_l2_pct` | **14.69** | 42.15 |
| `wall_shear_x_rel_l2_pct` | **12.36** | 27.59 |
| `wall_shear_y_rel_l2_pct` | **17.77** | 40.41 |
| `wall_shear_z_rel_l2_pct` | **19.15** | 83.23 |
| `volume_pressure_rel_l2_pct` | 7.60 | **6.98** |

### Per-epoch `val_primary` trajectory

`noproj` (o63hdjh6): improving each epoch through the full training budget.

| step | epoch | abupt | p_s | τ | τ_x | τ_y | τ_z | p_v |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 10883 | 1 | 18.638 | 12.827 | 20.956 | 17.618 | 25.775 | 26.784 | 10.186 |
| 21767 | 2 | 14.432 | 9.761 | 16.230 | 13.657 | 19.564 | 21.272 | 7.906 |
| 32651 | 3 | **13.095** | 8.592 | 14.688 | 12.359 | 17.769 | 19.154 | 7.599 |
| 33305 | partial 4¹ | 33.976 | 22.734 | 36.043 | 30.315 | 44.797 | 45.338 | 26.694 |

`proj` (xbbhtbwq): wall-shear axes are catastrophically bad from epoch 1 and never recover.

| step | epoch | abupt | p_s | τ | τ_x | τ_y | τ_z | p_v |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 10883 | 1 | 36.280 | 12.866 | 44.269 | 31.066 | 48.298 | 78.805 | 10.367 |
| 21767 | 2 | 33.745 | 9.433 | 42.285 | 28.610 | 42.745 | 80.201 | 7.733 |
| 32651 | 3 | **33.208** | 7.832 | 42.153 | 27.594 | 40.410 | 83.227 | 6.976 |
| 33284 | partial 4¹ | 33.226 | 7.690 | 42.246 | 27.520 | 40.359 | 83.729 | 6.832 |

¹ partial-epoch validation triggered by the train-timeout code-path (`timeout_hit and n_batches > 0`).
For `noproj` it's a 654-step partial epoch evaluated under the live cosine-EMA shadow before the
full eval; for `proj` it's effectively the same level as the converged model.

### `train/wallshear_pred_normal_rms` trajectory (proj arm only)

The projection loss applies MSE to the projected pair (pred and target both flattened to the
tangent plane), but **does not penalize the predicted normal component**. Empirically the normal
RMS overshoots and stays around 1.2 throughout training — the pred wall-shear vector has a large
component pointing along the local surface normal that contributes to the full-vector
`wall_shear_*_rel_l2_pct` metrics but is invisible to the loss.

| step | normal_rms |
|---:|---:|
|   158 | 0.758 |
|  1325 | 1.278 |
|  4955 | 1.182 |
|  9763 | 1.160 |
| 14303 | 1.201 |
| 20502 | 1.214 |
| 26139 | 1.239 |
| 33019 | 1.245 |

(80 samples downsampled across 3 epochs; the trajectory is essentially flat at ~1.2 after the
first few hundred steps. Full per-step series in W&B run `xbbhtbwq`.)

### W&B run IDs and URLs

| Arm | name | run id | url |
|---|---|---|---|
| control | `gilbert-6l-noproj` | `o63hdjh6` | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/o63hdjh6 |
| projection | `gilbert-6l-proj` | `xbbhtbwq` | https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/xbbhtbwq |

Both runs grouped under `wandb-group=gilbert-6l-tangential-sweep`. Total runtime 272.4 min each.
Peak GPU memory ≈ 78 GB on a single H100 (76% of 97 GB).

### Exact `train.py` commands

```bash
# control
CUDA_VISIBLE_DEVICES=0 python train.py \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --clip-grad-norm 1.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group gilbert-6l-tangential-sweep --wandb-name gilbert-6l-noproj \
  --agent gilbert

# projection
CUDA_VISIBLE_DEVICES=1 python train.py \
  --use-tangential-wallshear-loss \
  --model-layers 6 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --ema-decay-start 0.99 --ema-decay-end 0.9999 \
  --clip-grad-norm 1.0 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group gilbert-6l-tangential-sweep --wandb-name gilbert-6l-proj \
  --agent gilbert
```

## What happened

**The projection loss (in its current form) is decisively bad on the 6L base — it does not push
the wall-shear axes lower, it makes them dramatically worse.** There is no train→val divergence
at epoch 3+ to look for: divergence is already maximal at epoch 1.

The mechanism is the missing normal-component penalty (`target/train.py:1166-1236`). The
loss flattens both `pred` and `target` onto the local tangent plane, then takes MSE between the
projected pair. The predicted normal component never enters the loss, so it has no gradient
signal pulling it to zero. Validation/test, however, score the full 3-vector wall-shear
prediction — so the unrestrained normal-component drift (RMS ≈ 1.2 in normalized units, vs
target ≈ 0) shows up directly in `wall_shear_*_rel_l2_pct`. The effect is largest on `τ_z` (80.73
vs 18.54 control) because the body is mostly horizontal and `n_z` dominates the normal direction
for most surface points.

Compounding clip=1.0 with the projection loss does not fix this — the gradient that would shrink
the normal component does not exist at all in the projected MSE; clipping only bounds the
magnitude of whatever gradients the loss does produce.

Side note: surface pressure (7.75 vs 8.49) and volume pressure (13.60 vs 14.26) are slightly
better in the projection arm. With less of the surface-head capacity being burned trying to fit
the (apparently lossy) wall-shear-vector targets, the shared backbone has a bit more room for the
scalar channels. This is consistent with the head/loss-balance story but the gain is dwarfed by
the wall-shear regression. **Net composite (`abupt_axis_mean`) is +19.8 pp, a clear loss.**

The control arm itself underperforms the existing 6L baseline on `yi` (PR #14, `et4ajeqj`)
slightly: 14.28 vs 13.15 on `abupt_axis_mean`, with all wall-shear axes ~0.9–1.8 pp worse. The
delta versus PR #14 is the EMA schedule change in this PR's instructions
(`--ema-decay-start 0.99 --ema-decay-end 0.9999` cosine ramp here vs `--ema-decay 0.9995`
constant in PR #14). Three-epoch budgets are short, so the cosine ramp didn't finish ramping,
and the EMA shadow probably tracks the live model less tightly at the end of training. This is
**not** a robust new baseline; PR #14's recipe remains the better control.

### Verdict

**Projection loss does not help on the 6L base.** With the current implementation,
`--use-tangential-wallshear-loss` should not be combined with longer schedules or deeper bases.
The projection idea is salvageable but requires a code change: an explicit normal-component
penalty `λ · (n̂ · ŷ_pred)²` added to the surface loss. Given the strict
"only `target/train.py` per PR" constraint and the negative result, I'm reporting and stopping
here rather than pivoting to that code change inside this PR.

## Suggested follow-ups

1. **(Highest leverage)** New PR adding a normal-component penalty to the projection loss:
   `loss += lambda_n * (normal_dot.square()[mask].mean())`. Sweep `lambda_n ∈ {0.1, 1.0, 10.0}`
   on the 6L base. Without this, the projection loss appears strictly harmful.
2. **Drop the cosine EMA ramp for 3-epoch runs.** This PR's `--ema-decay-start 0.99
   --ema-decay-end 0.9999` underperformed PR #14's `--ema-decay 0.9995` on the same 6L base by
   ~1.1 pp `abupt`. The cosine ramp probably needs more epochs to be useful; a longer-budget
   sweep or a constant 0.9995 default for short runs is worth a confirmation experiment.
3. **More epochs on the 6L base.** PR #14 also stopped while still descending. The largest
   remaining gap is `τ_y` and `τ_z` (~4.5× AB-UPT). A longer 6L confirmation run (e.g. 6 epochs)
   would tell us how much of the gap is convergence-limited vs capacity-limited before chasing
   architectural changes.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #61 because The tangential projection arm (without normal penalty) completely catastrophic — abupt=34.07, which is 2.7× the baseline of 12.74. The projection removes gradient signal in the normal direction, but without a normal-suppression penalty the model has no incentive to suppress the predicted normal component, which grows unbounded. The normal-projection approach already had a second chance in PR #21 kohaku (normal suppression + projection on 6L) which scored abupt=15.93 best arm — also worse than baseline. Closing both lines of the tangential projection research. The physics-motivated approach is interesting but requires more careful formulation; revisit only if a fundamentally different implementation is available. — ADVISOR.
